### PR TITLE
Validate datasets versions

### PR DESCRIPTION
**Pull request by ElenaKhaustova**
## Description
Solves https://github.com/kedro-org/kedro/issues/4327

Merge before https://github.com/kedro-org/kedro/issues/4329

## Development notes

Added `_validate_versions` function to ensure all datasets in a catalog adhere to a versioning scheme - we allow single load version per dataset in the catalog and one save version for all datasets in the catalog. The function automatically updates the provided load versions based on the versions specified for the individual datasets. It also ensures all versioned datasets in the catalog share the same save version. If a conflict arises, a `VersionAlreadyExistsError` is raised.

Validation is applied to both `DataCatalog` and `KedroDataCatalog` when a catalog is created or the dataset is added.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/io/core.py [hunk 1]
```diff
@@ -955,3 +964,56 @@ def confirm(self, name: str) -> None:
...
+def _validate_versions(
+    datasets: dict[str, AbstractDataset] | None,
+    load_versions: dict[str, str],
+    save_version: str | None,
+) -> tuple[dict[str, str], str | None]:
+    """Validates and synchronizes dataset versions for loading and saving.
+
+    Insures consistency of dataset versions across a catalog, particularly
+    for versioned datasets. It updates load versions and validates that all
+    save versions are consistent.
```
Comment (merelcht) [inline kedro/io/core.py:978]:
```suggestion
    """Validates and synchronises dataset versions for loading and saving.

    Ensures consistency of dataset versions across a catalog, particularly
    for versioned datasets. It updates load versions and validates that all
    save versions are consistent.
```

File: kedro/io/core.py [hunk 2]
```diff
@@ -955,3 +964,56 @@ def confirm(self, name: str) -> None:
...
+
+    Args:
+        datasets: A dictionary mapping dataset names to their instances.
+            if None, no validation occurs.
+        load_versions: A mapping between dataset names and versions
+            to load.
+        save_version: Version string to be used for ``save`` operations
+            by all datasets with enabled versioning.
```
Comment (merelcht) [inline kedro/io/core.py:986]:
```suggestion
        save_version: Version string to be used for ``save`` operations
            by all datasets with versioning enabled.
```

File: kedro/io/core.py [hunk 3]
```diff
@@ -955,3 +964,56 @@ def confirm(self, name: str) -> None:
...
+        if isinstance(cur_ds, AbstractVersionedDataset) and cur_ds._version:
+            if cur_ds._version.load:
+                cur_load_versions[ds_name] = cur_ds._version.load
+            if cur_ds._version.save:
+                cur_save_version = cur_save_version or cur_ds._version.save
+                if cur_save_version != cur_ds._version.save:
+                    raise VersionAlreadyExistsError(
+                        f"Cannot add a dataset `{ds_name}` with `{cur_ds._version.save}` save version. "
+                        f"Save version set for the catalog is `{cur_save_version}`"
+                        f"All datasets in the catalog must have the same Save version."
```
Comment (merelcht) [inline kedro/io/core.py:1016]:
```suggestion
                        f"All datasets in the catalog must have the same save version."
```

**Review by merelcht — APPROVED** [review]
LGTM, and all makes sense 👍 

One thing I was wondering is if this could somehow break existing workflows that use the `DataCatalog`?

**Comment by ElenaKhaustova** [comment]
> LGTM, and all makes sense 👍
> 
> One thing I was wondering is if this could somehow break existing workflows that use the `DataCatalog`?

That's a good question. Technically, we always required `DataCatalog` with versioned datasets, and we have a note in the docstrings that the catalog can have only one version. But we never validated it, and this requirement can be bypassed. So we consider this change a fix—that's why we applied it for both old and new catalogs. We also think this shouldn't be the common case, but even if some workflows break, that's because people used versioning not as expected.

The alternative was to apply these changes just to `KedroDataCatalog`.

**Comment by merelcht** [comment]
> That's a good question. Technically, we always required `DataCatalog` with versioned datasets to have only one save version. But we never validated it, and this requirement can be bypassed. So we consider this change as a fix—that's why we applied it for both old and new catalogs. We also think this shouldn't be the common case, but even if some workflows break, that's because people used versioning not as expected.
> 
> The alternative is to apply these changes just to `KedroDataCatalog`.

Yes so technically speaking this is breaking for people who have been passing instantiated dataset objects to catalog constructor with different save versions. It's a very rare thing to do and I agree this is actually improving the expected behaviour. So let's just keep it like this, but then at least if someone does complain we know what's going on.

**Review by idanov — APPROVED** [review]
LGTM 👍

**Pull request merged**